### PR TITLE
LBM

### PR DESCRIPTION
**Pull request by AliKarakus**
DG / Lattice-Boltzmann solver with basic functionality.

### CODE REVIEW

**Review by tcew — COMMENTED** [review]
LGTM. 

Can you elaborate a bit on the difference to bns.

**Comment by AliKarakus** [comment]
> LGTM.
> 
> Can you elaborate a bit on the difference to bns.

Both solvers govern the same physics and expecting to achieve similar results for low Mach flows. I think there are couple points of using DG- LBM, ​
--> it is very straight-forward to extend to different physics i.e. heat transfer, natural convection, multi-phase (I am working on connecting with Level-Set method for sharp interfaces) and multi-relaxation. 
​--> comparing  with BNS especially in boundary layers.
--> possible publications with extending to deforming meshes, pml, multi-phase, etc.

File: solvers/lbs/setups/setupQuad2D.rc [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+LSERK4
+
+[CFL NUMBER]
+0.5
+
+[START TIME]
+0
+
+[FINAL TIME]
+0.1
```
Comment (noelchalmers) [inline solvers/lbs/setups/setupQuad2D.rc:53]:
```suggestion
10.0
```

File: solvers/lbs/setups/setupTet3D.rc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+LSERK4
+
+[CFL NUMBER]
+0.5
+
+[START TIME]
+0
+
+[FINAL TIME]
+0.1
```
Comment (noelchalmers) [inline solvers/lbs/setups/setupTet3D.rc:56]:
```suggestion
10.0
```

File: solvers/lbs/setups/setupTri2D.rc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+LSERK4
+
+[CFL NUMBER]
+0.5
+
+[START TIME]
+0
+
+[FINAL TIME]
+0.1
```
Comment (noelchalmers) [inline solvers/lbs/setups/setupTri2D.rc:53]:
```suggestion
10.0
```

File: test/testLbs.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+from test import *
+
+lbsData2D = lbsDir + "/data/lbsGaussian2D.h"
+lbsData3D = lbsDir + "/data/lbsGaussian3D.h"
+
+def lbsSettings(rcformat="2.0", data_file=lbsData2D,
+               mesh="BOX", dim=2, element=4, nx=10, ny=10, nz=10, boundary_flag=1,
+               degree=4, thread_model=device, platform_number=0, device_number=0,
+               viscosity=0.01, velmodel="D2Q9", time_integrator="LSERK4", cfl=0.5, start_time=0.0, final_time=0.1,
```
Comment (noelchalmers) [inline test/testLbs.py:37]:
```suggestion
               viscosity=0.01, velmodel="D2Q9", time_integrator="LSERK4", cfl=0.5, start_time=0.0, final_time=10.0,
```

**Review by noelchalmers — CHANGES_REQUESTED** [review]
This looks really good to me. A couple questions/comments:

- I think it's a good idea to make the default final time t=10 in the setup.rcs so that the runs produce several frames. 
- Same comment for the testing runs. Otherwise only a small number of timesteps are actually performed. (Changing this will require updated 'golden' norms).
- I suspect there is a subtle bug in the Quad2D code path. Running the default setup to t=10 doesn't pass the 'eyeball' norm. The Gaussian doesn't seem to move at all.

I see a several places where early PML and/or Multirate support is enabled/commented. @AliKarakus  is the plan to add these features in the future? They're not hurting anything so no worries either way. 

I also did some quick profiling on large problem setups. Probably unsurprisingly, several kernels are register heavy, the volume and phaseField kernels in particular are often at <30% peak DRAM BW. Not a problem, just an FYI. 

I second @tcew 's comment. I'm not particularly familiar with LBM, so it would be cool to see the full formulation written somewhere!

Thanks @AliKarakus !

**Comment by noelchalmers** [comment]
@AliKarakus @tcew These test runs are going to be bitten by the latest OCCA changes complaining about `@barrier("local")`. Let ignore these test fails for the moment. 

Should we make occa a git submodule to avoid this kind of thing in the future?

**Comment by tcew** [comment]
> Should we make occa a git submodule to avoid this kind of thing in the future?

Yes. I think this is a very good and timely idea. 

I anticipate quite a few changes in occa and we do not want to be continually held up.

**Comment by AliKarakus** [comment]
@noelchalmers Thanks for the feedback, I double checked  all the tests on my local machine. Everything works smoothly. 

--> I see a several places where early PML and/or Multirate support is enabled/commented. @AliKarakus is the plan to add these features in the future? They're not hurting anything so no worries either way.

I am going to work on a PML formulation, but multi-rate time-stepping  creates some issues because of the time-step size dependent properties in LBM. Need to check details a little more.  Currently my target is to combine this basic DG-LBM formulation with Level-Set for bubble dynamics and contact line problems. 

--->I also did some quick profiling on large problem setups. Probably unsurprisingly, several kernels are register heavy, the volume and phaseField kernels in particular are often at <30% peak DRAM BW. Not a problem, just an FYI.

Yes, I think  this problem is arising from the high number of fields i.e. 9 for 2D and 15 or 19 for 3D.  At the moment, I don't know how to overcome this performance issue quickly  as BNS outperforms LBM even in 2D test cases. 

--->I'm not particularly familiar with LBM, so it would be cool to see the full formulation written somewhere!
It will be available soon (hopefully)  with DG-LBM/LS paper. 

Thanks a lot @noelchalmers and  @tcew .

**Comment by noelchalmers** [comment]
@AliKarakus Can you merge the latest master branch? That should fix our OCCA issues and re-run the CI checks.

**Comment by tcew** [comment]
@AliKarakus @noelchalmers - the Lbs tests seem to take quite a while. In particular the tet 3d case seems to take 20 minutes unless I misread the test report.

**Review by noelchalmers — APPROVED** [review]

**Comment by noelchalmers** [comment]
LGTM. The whole suite of tests take around 24 minutes.

**Comment by tcew** [comment]
> LGTM. The whole suite of tests take around 24 minutes.

@noelchalmers would you like to do the honors and merge this at your convenience ?

**Comment by tcew** [comment]
Great work @AliKarakus and thanks to @noelchalmers for overseeing the merge !

**Comment by AliKarakus** [comment]
Thanks @tcew and @noelchalmers. I am deleting the lbm branch now.

**Pull request merged**